plugins/kubernetes/app/models/watchers/deploy_watcher.rb

```diff
@@ -0,0 +1,83 @@
+require 'celluloid'
+require 'celluloid/autostart'
+
+module Watchers
+  # Instantiated when a Kubernetes deploy is created to watch the status
+  class DeployWatcher
+    include Celluloid
+    include Celluloid::Notifications
+
+    finalizer :on_termination
+
+    def initialize(release)
+      @release = release
+      @current_rcs = {}
+      Rails.logger.info "Start watching K8s deploy: #{@release}"
+    end
+
+    def watch
+      @release.release_docs.each do |release_doc|
+        subscribe "#{release_doc.replication_controller_name}", :handle_update
+      end
+    end
+
+    def handle_update(topic, data)
+      release_doc = release_doc_from_rc_name(topic)
+      pod = Events::Pod.new(data)
+      return Rails.logger.error('invalid k8s pod event') unless pod.valid?
+      update_replica_count(release_doc, pod)
+      send_event(role: release_doc.kubernetes_role.name,
+                 deploy_group: release_doc.deploy_group.name,
+                 target_replicas: release_doc.replica_target,
+                 live_replicas: release_doc.replicas_live)
+      end_deploy if deploy_finished?
+    end
+
+    private
+
+    def deploy_finished?
+      @release.release_docs.all?(&:live?)
+    end
+
+    def release_doc_from_rc_name(name)
+      @release.release_docs.select { |doc| doc.replication_controller_name == name }.first
+    end
+
+    def on_termination
+      send_event(msg: 'Finished Watching Deploy!')
```

Comment:
It would be cleaner to also unsubscribe from notifications. I think we might see some stray messages in the logs in some edge cases.